src/vs/workbench/contrib/terminal/browser/terminalInstance.ts

```diff
@@ -985,6 +986,9 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
 	}
 
 	private _createProcess(): void {
+		if (this._xtermTypeAhead) {
+			this._processManager.onBeforeProcessData(e => this._xtermTypeAhead?.onBeforeProcessData(e));
```

Comment:
Right now it's nice that the only thing inside TerminalInstance relating to type ahead is that it loads the addon, everything else is encapsulated within the addon. A slight improvement to this I think is to create a `this._xtermTypeAhead.setProcessManager` which gets called inside `reuseTerminal`, a little more explicit about why it's there and lets the addon handling the events.

Reply:
I actually suggested going this way to Tyler initially. Before my mistaken assumption was that the process manager's lifetime was equivalent to the lifetime of the terminal instance, so I had it injected into the constructor of the addon on which we'd set up listeners. Since it's not actually the case, I thought a better approach would be to make the type ahead addon unaware of the process manager (or generally the way in which pty input is received) and be called into. Lessens the amount of knowledge it carries slightly.

Reply:
My incentives are generally pushing knowledge down into sub components to make `TerminalInstance` smaller/simpler😉, self contained addons that manage everything are great. Feel free to leave this as is, the ideal fix would probably be some event on process manager and making it aware of multiple processes or something. `TerminalInstance.reuseTerminal` and what it does is pretty ugly currently.

Reply:
So I realized I needed some sort of `reset()` method on the AddOn because the `lastRow` needed to be wiped out otherwise the bad behavior would show up after a `reuseTerminal()` if you had something typed on the line.

Since I needed a `reset` anyway, I have it passing in the new reference of `processManager` doing more what @Tyriar suggested here.